zookeeper-server/src/main/java/org/apache/zookeeper/server/ZooKeeperServer.java

```diff
@@ -1383,4 +1392,85 @@ public void setResponseCachingEnabled(boolean isEnabled) {
     public ResponseCache getReadResponseCache() {
         return isResponseCachingEnabled ? readResponseCache : null;
     }
+
+    protected void registerMetrics() {
```

Comment:
Probably move this logic to ServerMetrics, and call ServerMetrics.registerMetrics(this) in this class, so we have a centralized way to manage register and unregister.

Reply:
@lvfangmin  we should leverage Polymorphism, every subclass register his own metrics + the metrics of its parent class

Reply:
@lvfangmin if I move this to ServerMetrics we will have a lot of "instanceof" like the old code

Reply:
That makes sense to me, are we seeing this pattern often, should we define an interface for these methods?

Reply:
By now it is already the hierarchy of ZooKeeperServer.
I am not sure it is worth to add another interface.
If you feel strong I will do it